### DISCUSSION TITLE
Contour Lines Unstructured Grids

### DISCUSSION THREAD

**Discussion by AlexanderJuestel** [pyvista/pyvista#4754] [Q&A] [answered]
Hello,

I was wondering if there is a way to create contour lines for UnstructuredGrids? 

In particular, I have a grid that represents the subsurface temperature (see image). The temperature increases with depth fairly homogeneously. Is it now possible to add contour lines similar to PolyData to first of all show the depth intervals and secondly the temperatures in the subsurface?

![image](https://github.com/pyvista/pyvista/assets/45469915/37b47acd-9035-4566-8ed9-9de1b26dd7c7)

**Comment by MatthewFlamm** [comment]
You can still use contour filter for `UnstructuredGrid`.  Does your `UnstructuredGrid` have 3D cells or 2D cells like `PolyData`?  You can still use contour on `UnstructuredGrid`, but it will give you a 2D surface when using 3D cells.  This might be useful to you if you wanted to look inside the domain for the isosurfaces using transparency for the whole mesh, for example .

If you only want to plot the exterior surfaces here, can you first `extract_surface` to get a `PolyData`, then use `contour` on only that surface?

**Reply by AlexanderJuestel** [reply]
@MatthewFlamm, thanks for your quick response. When using `contour()` on the grid, I get the following error. I also get it when specifying the scalars and set `preference='cell'`. 

![image](https://github.com/pyvista/pyvista/assets/45469915/0a8f3f96-c2aa-44d9-ad13-1e52e0e9fabf)


Maybe a little more context: 

I have a GeoDataFrame consisting of 838750 X, Y, and Z points with associated temperature values. 
![image](https://github.com/pyvista/pyvista/assets/45469915/85f8b21e-515b-4974-849f-746ea0933e38)

I used PVGeo to create the grid: 

```python

vtkpoints = PVGeo.points_to_poly_data(data_gdf_clipped[['X', 'Y', 'Z']].values)
vtkpoints['Temperature'] = data_gdf_clipped['Untergrundtemperatur'].values

voxelizer = PVGeo.filters.VoxelizePoints()
grid = voxelizer.apply(vtkpoints)
grid['Temperature'] = data_gdf_clipped['Untergrundtemperatur'].values
grid
```

![image](https://github.com/pyvista/pyvista/assets/45469915/4d7b3ec1-3d26-4468-9055-11d1eebadeeb)

Maybe there is an easier way of creating a grid in the first place but that has worked for me so far.

**Reply by MatthewFlamm** [reply]
The contour filter only works on point data and you only have cell data.  As far as I know, this is true for all `pyvista.DataSet` instances including `PolyData` and `UnstructuredGrid`.  You can use `cell_data_to_point_data` to get point data for contouring.

EDIT: I don't know anything about PVGeo, so not sure if it is possible or better to get the point data directly.

**Reply by AlexanderJuestel** [reply]
It gets more and more interesting. Casting the cell-to-point data using `grid.cell_data_to_point_data()` results in the following. The contours are extracted correctly but the input is already a little off. (blue an red colors are the temperatures, colors with colormap viridis the contours)

![image](https://github.com/pyvista/pyvista/assets/45469915/a26d39b4-f5a6-483d-8864-5beb25c52171)

**Reply by AlexanderJuestel** [reply, marked as answer]
I got it! I think what did the trick was passing the range of the temperature array.....

```python
grid_point_data = grid.cell_data_to_point_data()
contours = grid_point_data.contour(np.arange(0, 190, 20))
```

![image](https://github.com/pyvista/pyvista/assets/45469915/649aaa48-8967-453e-96d6-c84852192e99)

**Reply by MatthewFlamm** [reply]
Looks great!

Thanks for coming back with the ultimate solution to this, it will be helpful to others having the same issues. From my understanding, the keys were getting point data and using the appropriate scale for the contour isovalues.
